File: iqchannels-sdk/src/main/java/ru/iqchannels/sdk/lib/InternalIO.java
/*
 * Copyright (c) 2017 iqstore.ru.
 * All rights reserved.
 */

package ru.iqchannels.sdk.lib;

import androidx.annotation.Nullable;

import java.io.BufferedInputStream;
import java.io.BufferedOutputStream;
import java.io.ByteArrayInputStream;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;

/**
 * Created by Ivan Korobkov i.korobkov@iqstore.ru on 27/04/2017.
 */

public class InternalIO {
    public static interface ProgressCallback {
        void onProgress(int progress);
    }

    private InternalIO() {}

    public static void copy(File src, File dst) throws IOException {
        InputStream in = new FileInputStream(src);
        try {
            OutputStream out = new FileOutputStream(dst);
            try {
                copy(in, out);
            } finally {
                out.close();
            }
        } finally {
            in.close();
        }
    }

    public static void copy(byte[] src, OutputStream dst) throws IOException {
        copy(src, dst, null);
    }

    public static void copy(byte[] src, OutputStream dst,
                            @Nullable ProgressCallback callback) throws IOException {
        ByteArrayInputStream input = new ByteArrayInputStream(src);

        int n;
        int total = 0;
        int progress = 0;

        byte[] buf = new byte[16 * 1024];
        while ((n = input.read(buf)) > -1) {
            dst.write(buf, 0, n);
            dst.flush();

            total += n;
            int newProgress = (total * 100) / src.length;
            if (newProgress != progress) {
                progress = newProgress;
                if (callback != null) {
                    callback.onProgress(progress);
                }
            }
        }
    }

    public static void copy(InputStream src, OutputStream dst) throws IOException {
        int n;

        byte[] buf = new byte[16 * 1024];
        while ((n = src.read(buf)) > -1) {
            dst.write(buf, 0, n);
            dst.flush();
        }
    }
}

File: iqchannels-sdk/src/main/java/ru/iqchannels/sdk/lib/InternalIO.kt
/*
 * Copyright (c) 2017 iqstore.ru.
 * All rights reserved.
 */
package ru.iqchannels.sdk.lib

import java.io.ByteArrayInputStream
import java.io.File
import java.io.FileInputStream
import java.io.FileOutputStream
import java.io.IOException
import java.io.InputStream
import java.io.OutputStream

object InternalIO {
	@Throws(IOException::class)
	fun copy(src: File?, dst: File?) {
		val `in`: InputStream = FileInputStream(src)
		try {
			val out: OutputStream = FileOutputStream(dst)
			try {
				copy(`in`, out)
			} finally {
				out.close()
			}
		} finally {
			`in`.close()
		}
	}

	@JvmOverloads
	@Throws(IOException::class)
	fun copy(
		src: ByteArray,
		dst: OutputStream,
		callback: ProgressCallback? = null
	) {
		val input = ByteArrayInputStream(src)
		var n: Int
		var total = 0
		var progress = 0
		val buf = ByteArray(16 * 1024)
		while (input.read(buf).also { n = it } > -1) {
			dst.write(buf, 0, n)
			dst.flush()
			total += n
			val newProgress = total * 100 / src.size
			if (newProgress != progress) {
				progress = newProgress
				callback?.onProgress(progress)
			}
		}
	}

	@Throws(IOException::class)
	fun copy(src: InputStream, dst: OutputStream) {
		var n: Int
		val buf = ByteArray(16 * 1024)
		while (src.read(buf).also { n = it } > -1) {
			dst.write(buf, 0, n)
			dst.flush()
		}
	}

	interface ProgressCallback {
		fun onProgress(progress: Int)
	}
}
